www/blog/2023-01-08-tinyrpc-client.mdx

```diff
@@ -0,0 +1,210 @@
+
+## Implementing a tiny tRPC client
+
+We'll first create a helper function for creating recursive proxies - `createRecursiveProxy`:
+
+```ts
+interface ProxyCallbackOptions {
+  path: string[];
+  args: unknown[];
+}
+
+type ProxyCallback = (opts: ProxyCallbackOptions) => unknown;
+
+const noop = () => {
+  // noop
+};
+
+function createRecursiveProxy(callback: ProxyCallback, path: string[]) {
+  const proxy: unknown = new Proxy(noop, {
+    get(_obj, key) {
+      if (typeof key !== 'string' || key === 'then') {
+        // special case for if the proxy is accidentally treated
+        // like a PromiseLike (like in `Promise.resolve(proxy)`)
+        return undefined;
+      }
+      return createRecursiveProxy(callback, [...path, key]);
+    },
+    apply(_1, _2, args) {
+      return callback({
+        args,
+        path,
+      });
+    },
+  });
+
+  return proxy;
+}
+```
+
+This looks a bit magical, what does this do?
+
+- The `get` method handles property accesses such as `post.byId`, and the `apply` method handles function calls such as `.query(args)`. The key is the property name we're accessing, so when we type `post` our `key` will be `post`, and when we type `post.byId` our `key` will be `byId`. The recursive proxy combines all of these keys into a final path, e.g. ["post", "byId", "query"], that we can use to determine the URL we want to call on the server.
+- The `apply` method is called when we call a function on the proxy, such as `.query(args)`. The `args` is the arguments we pass to the function, so when we call `post.byId.query(args)` our `args` will be our input, which we'll provide as query parameters or request body. We'll provide a `callback` function to the `createRecursiveProxy` function that will be called with the path and args. This callback will be a `fetch` request to our server.
+
+:::info
+
+The `noop` here comes from the fact that we dont have a client side object we want to remap to. We only care about the path we're accessing and then we'll use that to call our server.
+
+:::
+
+Now that we have this helper and knows what it does, let's use it to create our client. We'll provide the `createRecursiveProxy` a callback that will take the path and args and request the server using `fetch`.
+
+```ts
+export const createTinyRPCClient = (url: string) =>
+  createRecursiveProxy(async (opts) => {
+    const parts = [...opts.path]; // ["post", "byId", "query"]
+    const method = parts.pop() as 'query' | 'mutate';
+    const path = parts.join('.'); // "post.byId"
+
+    let uri = `${url}/${path}`;
+    const input = opts.args[0];
+    const stringifiedInput = input !== undefined && JSON.stringify(input);
+    let requestBody: undefined | string = undefined;
+    if (stringifiedInput !== false) {
+      if (last === 'query') {
+        uri += `?input=${encodeURIComponent(stringifiedInput)}`;
+      } else {
+        body = stringifiedInput;
+      }
+    }
+
+    const json: TRPCResponse = await fetch(uri, {
```

Comment:
"What is `TRPCResponse`"

Reply:
Link to https://trpc.io/docs/rpc when explaining this maybe :)

Reply:
Not quite sure what I should have written here tbf...

![CleanShot 2023-01-13 at 00 55 54](https://user-images.githubusercontent.com/51714798/212205869-46fcbce9-ba2c-4ab5-8aad-cc2aa1697917.png)